superqt/combobox/_enum_combobox.py

```diff
@@ -0,0 +1,118 @@
+from enum import Enum
+from typing import Optional, Type, TypeVar, Union
+
+from typing_extensions import get_args, get_origin
+
+from ..qtcompat.QtCore import Signal
+from ..qtcompat.QtWidgets import QComboBox
+
+EnumType = TypeVar("EnumType", bound=Enum)
+
+
+NONE_STRING = "----"
+
+
+def _get_name(enum_value: Enum):
+    """Create human readable name if user does not provide own implementation of __str__"""
+    if enum_value.__str__.__module__ != "enum":
+        # check if function was overloaded
+        name = str(enum_value)
+    else:
+        name = enum_value.name.replace("_", " ")
+    return name
+
+
+class QEnumComboBox(QComboBox):
+    """
+    ComboBox presenting options from a python Enum.
+
+    If the Enum class does not implement `__str__` then a human readable name
+    is created from the name of the enum member, replacing underscores with spaces.
+    """
+
+    currentEnumChanged = Signal(object)
+
+    def __init__(self, parent=None, enum_class: Optional[Type[EnumType]] = None):
+        super().__init__(parent)
+        self._enum_class = None
+        self._optional = False
+        if enum_class is not None:
+            self.setEnumClass(enum_class)
+        self.currentIndexChanged.connect(self._emit_signal)
+
+    def setEnumClass(self, enum: Type[EnumType]):
+        """
+        Set enum class from which members value should be selected
+        """
+        self.clear()
+        if get_origin(enum) is Union:
```

Comment:
hmm.  while i think this sort of type annotation introspection makes sense in the context of magicgui (where we're deriving settings from function signatures), it feels a little out of place here.  What would you think about just have another parameter `nullable`, or `allow_none`, or `optional` or something...

```py
def setEnumClass(self, enum: Type[EnumType], nullable=False): ...
```

also, now that I'm looking at typing, I'm not sure the `TypeVar` is working as intended (as opposed to just using `EnumMeta`).  If the goal is to have it know that when you provide a specific `Enum` class, that's what you'll get out of `enumClass()`, or that's what is needed in `setCurrentEnum`, then I'm not sure it works that way (since we're relying on a runtime argument).  Thoughts?

Reply:
> hmm. while i think this sort of type annotation introspection makes sense in the context of magicgui (where we're deriving settings from function signatures), it feels a little out of place here. What would you think about just have another parameter `nullable`, or `allow_none`, or `optional` or something...

changed


> also, now that I'm looking at typing, I'm not sure the `TypeVar` is working as intended (as opposed to just using `EnumMeta`). If the goal is to have it know that when you provide a specific `Enum` class, that's what you'll get out of `enumClass()`, or that's what is needed in `setCurrentEnum`, then I'm not sure it works that way (since we're relying on a runtime argument). Thoughts?

I do not remember why I use `TypeVar` when I implement this in PartSeg. And I use PartSeg implementation as a base for this. It was connected with type checking, but maybe it should be changed to EnumMeta. I will try and see if there will be some warnings.